Add Flatpak Docs (#2279)
tauri-apps/tauri-docs: src/content/docs/distribute/flatpak.mdx

---
title: Flathub
sidebar:
  order: 1
---

import { Tabs, TabItem, Card } from '@astrojs/starlight/components'

For detailed information on how Flatpak works, you can read [Building your first Flatpak](https://docs.flatpak.org/en/latest/first-build.html)

This guide assumes you want to distribute your Flatpak via [Flathub](https://flathub.org/), the most commonly used platform for Flatpak distribution. If you plan on using other platforms, please consult their documentation instead.
### Prerequisites

To test your app inside the Flatpak runtime you can build the Flatpak locally first before uploading your app to Flathub. This can also be helpful if you want to quickly share development builds.

**1. Install `flatpak` and `flatpak-builder`**

To build Flatpaks locally you need the `flatpak` and `flatpak-builder` tools. For example on Ubuntu you can run this command:
<Tabs>
  <TabItem label="Debian">

```sh

> sudo apt update

Reviewer: can we remove the update commands again? I'd prefer to keep the install command clean/small. (Thinking about doing the same for the initial tauri reqs docs but note sure yet)

sudo apt install flatpak flatpak-builder
```

  </TabItem>
  <TabItem label="Arch">

```sh
sudo pacman -Syu
sudo pacman -S --needed flatpak flatpak-builder
```
  </TabItem>
  <TabItem label="Fedora">

```sh
sudo dnf check-update
sudo dnf install flatpak flatpak-builder
```
  </TabItem>
  <TabItem label="Gentoo">

```sh
sudo emerge --ask \
sys-apps/flatpak \
dev-util/flatpak-builder
```
  </TabItem>
</Tabs>

**2. Install the Flatpak Runtime**

```shell
flatpak install flathub org.Gnome.Platform//46 org.Gnome.Sdk//46
```

> **3. [Build the .deb of your tauri-app](https://tauri.app/v1/guides/building/linux#debian)**

Reviewer: ```suggestion
**3. [Build the .deb of your tauri-app](https://tauri.app/v1/guides/building/linux#debian)**

{/* TODO: Update link to v2 */}
```
just so we track it

**4. Create the manifest**

```yaml
id: org.your.id

runtime: org.gnome.Platform
runtime-version: '46'
sdk: org.gnome.Sdk

command: tauri-app
finish-args:
  - --socket=wayland # Permission needed to show the window
  - --socket=fallback-x11 # Permission needed to show the window
  - --device=dri # OpenGL, not necessary for all projects
  - --share=ipc

modules:
  - name: binary
    buildsystem: simple
    sources:
      - type: file
        url: https://github.com/your_username/your_repository/releases/download/v1.0.1/yourapp_1.0.1_amd64.deb
        sha256: 08305b5521e2cf0622e084f2b8f7f31f8a989fc7f407a7050fa3649facd61469 # This is required if you are using a remote source

>         only-arches: [x86_64] #This allows you to make your source architecture specific so that arm users dont install amd64 apps. Another acceptable value is [aarch64].

Reviewer: ```suggestion
        only-arches: [x86_64] # This allows you to make your source architecture specific so that arm users dont install amd64 apps. Another acceptable value is [aarch64].
```

Any command with a `(host)` label can be executed on your host, while commands without this label must be executed inside the container.

> `(host)` commands must be executed from inside the flatpak directory on your host.

Reviewer: I wonder if we should remove all the container vs host talk and just assume everything's on the host to not add confusion 🤔

The Gnome 46 runtime includes all dependencies of the standard Tauri app with their correct versions.

> Assume the flatpak manifest is in `/flatpak/org.your.id.yml` and the tauri project is in `/tauri-app`.

Reviewer: > Assume the flatpak manifest is in `/flatpak/org.your.id.yml` and the tauri project is in `/tauri-app`.

I think this is still confusing, i don't even really get it myself tbh

This can introduce various bugs that can be very hard to find.
Therefore, it is recommended to build the library your program depends on from source inside the flatpak as a build step.

## Submitting to flathub

> ***1. Fork The [Flathub Repository]("https://github.com/flathub/flathub/fork")***

Reviewer: ```suggestion
***1. Fork The [Flathub Repository](https://github.com/flathub/flathub/fork)***
```
(it links to https://deploy-preview-2279--tauri-v2.netlify.app/distribute/flatpak/%22https://github.com/flathub/flathub/fork%22 right now)

***2. Clone the Fork***
```shell
git clone --branch=new-pr git@github.com:your_github_username/flathub.git
```

***3. Enter the repository***
```shell
cd flathub
```
***4. Create a new branch***
```shell
git checkout -b your_app_name
```

> ***5. Add your apps manifest to branch. Commit it, and then push it.***

Reviewer: ```suggestion
***5. Add your apps manifest to the branch. Commit your changes, and then push them.***
```

We should think about showing the git commands for that too.

***6. Open a pull request against the `new-pr` branch on github***

***7. Your app will now enter the review process in which you may be asked to make changes to your project.***

> When your pull request is approved then you will recieve an invitation to edit your apps repository. From here on you can update your app continuosly.

Reviewer: ```suggestion
When your pull request is approved then you will receive an invitation to edit your apps repository. From here on you can update your app continuously.
```

You can read more about this [in the flatpak documentation](https://docs.flatpak.org/en/latest/dependencies.html#bundling)

import Stub from '@components/Stub.astro';

> <Stub />

Reviewer: Stub can be removed now
